lib/puma/reactor.rb

```diff
@@ -1,405 +1,121 @@
+        timeout_at, _ = @ios.first
+        select(timeout_in(timeout_at))
 
-              break if @timeouts.empty?
-            end
+        timeout!
 
-            calculate_sleep
-          end
-        end
+        # Register any newly-added objects from the @input Queue.
+        register(@input.pop) until @input.empty?
       end
-    end
-
-    def clear_monitor(mon)
-      @selector.deregister mon.value
-      @monitors.delete mon
-    end
-
-    public
 
-    def run
-      run_internal
-    ensure
-      @trigger.close
-      @ready.close
+      # Wake up all remaining objects on shutdown.
+      @ios.each(&method(:wakeup!))
+    rescue StandardError => e
+      STDERR.puts "Error in reactor loop escaped: #{e.message} (#{e.class})"
+      STDERR.puts e.backtrace
+      retry
     end
 
-    def run_in_thread
-      @thread = Thread.new do
-        Puma.set_thread_name "reactor"
-        begin
-          run_internal
-        rescue StandardError => e
-          STDERR.puts "Error in reactor loop escaped: #{e.message} (#{e.class})"
-          STDERR.puts e.backtrace
-          retry
-        ensure
-          @trigger.close
-          @ready.close
-        end
+    # Monitors registered IO objects until the specified timeout,
+    # calling #wakeup! on any that become readable.
+    def select(timeout)
+      @selector.select(timeout) {|mon| wakeup! mon.value}
+    rescue IOError => e
+      Thread.current.purge_interrupt_queue if Thread.current.respond_to? :purge_interrupt_queue
+      # An IOError is raised if any IOs are closed, so wakeup all closed IOs in this case.
+      if (closed = @ios.select {|_, io| io.closed?}).any?
+        STDERR.puts "Error in select: #{e.message} (#{e.class})"
+        STDERR.puts e.backtrace
+        closed.each(&method(:wakeup!))
+        retry
+      else
+        raise
       end
     end
 
-    # The `calculate_sleep` sets the value that the `NIO::Selector#select` will
-    # sleep for in the main reactor loop when no sockets are being written to.
-    #
-    # The values kept in `@timeouts` are sorted so that the first timeout
-    # comes first in the array. When there are no timeouts the default timeout is used.
-    #
-    # Otherwise a sleep value is set that is the same as the amount of time it
-    # would take for the first element to time out.
-    #
-    # If that value is in the past, then a sleep value of zero is used.
-    def calculate_sleep
-      if @timeouts.empty?
-        @sleep_for = DefaultSleepFor
-      else
-        diff = @timeouts.first.value.timeout_at.to_f - Time.now.to_f
-
-        if diff < 0.0
-          @sleep_for = 0
-        else
-          @sleep_for = diff
-        end
-      end
+    # Wakeup all objects that have timed out.
+    def timeout!
+      # @ios is sorted by timeout so we can efficiently take only timed out objects.
+      @ios.take_while {|timeout_at, _| timeout_in(timeout_at) == 0}.each(&method(:wakeup!))
```

Comment:
This needs to be broken out into multiple lines and probably expressions. It's quite different to figure out at present.